### PR TITLE
Use patch helper from cluster-api

### PR DESCRIPTION
**Pull request by detiber**
**What this PR does / why we need it**:

Use the patch helper from cluster-api to avoid having to replicate correct patching semantics.

**Release note**:
```release-note
NONE
```

### CODE REVIEW

**Comment by chuckha** [comment]
did we already bump the version of capi to latest master?

File: controllers/kubeadmconfig_controller.go [hunk 1]
```diff
@@ -119,13 +120,18 @@ func (r *KubeadmConfigReconciler) Reconcile(req ctrl.Request) (_ ctrl.Result, re
...
-	// Store Config's state, pre-modifications, to allow patching
-	patchConfig := client.MergeFrom(config.DeepCopy())
+	// Initialize the patch helper
+	patchHelper, err := patch.NewHelper(config, r)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+	// Always attempt to Patch the KubeadmConfig object and status after each reconciliation.
 	defer func() {
-		err := r.patchConfig(ctx, config, patchConfig)
```
Comment (vincepri) [inline controllers/kubeadmconfig_controller.go]: Should we also remove the method from this Reconciler?

File: go.sum [hunk 2]
```diff
@@ -254,9 +254,14 @@ k8s.io/utils v0.0.0-20190607212802-c55fbcfc754a h1:2jUDc9gJja832Ftp+QbDV0tVhQHMI
...
 sigs.k8s.io/cluster-api v0.0.0-20190809134526-b8af96f0a42a h1:aDiczQxkpWlxe5yzqm6rJNe5orMuHTt7Q0MqDJOJ/bg=
 sigs.k8s.io/cluster-api v0.0.0-20190809134526-b8af96f0a42a/go.mod h1:CgxuxJy6W8onO8duP2ED+s0kwNxsNWIGEXMnSUsxGcw=
+sigs.k8s.io/cluster-api v0.0.0-20190820130448-9fd8e4cbea0f h1:gL8HQOSMGhMvL2BZNQSuKdXe02se34Uf29Rvq3iOfP8=
+sigs.k8s.io/cluster-api v0.0.0-20190820130448-9fd8e4cbea0f/go.mod h1:zBOsRI6nVKzD6CPmA/3qYpLCckc5FVxMBxhvBxuDNsg=
 sigs.k8s.io/controller-runtime v0.2.0-beta.5 h1:W2jTb239QEwQ+HejhTCF9GriFPy2zmo1I6pPmJTeEy8=
 sigs.k8s.io/controller-runtime v0.2.0-beta.5/go.mod h1:HweyYKQ8fBuzdu2bdaeBJvsFgAi/OqBBnrVGXcqKhME=
+sigs.k8s.io/controller-runtime v0.2.0-rc.0 h1:49JLOielmXfrd44Cmk2c0eeIkQ/Vq4AvfqsZqya16/E=
+sigs.k8s.io/controller-runtime v0.2.0-rc.0/go.mod h1:HweyYKQ8fBuzdu2bdaeBJvsFgAi/OqBBnrVGXcqKhME=
 sigs.k8s.io/controller-tools v0.2.0-beta.5/go.mod h1:8t/X+FVWvk6TaBcsa+UKUBbn7GMtvyBKX30SGl4em6Y=
+sigs.k8s.io/controller-tools v0.2.0-rc.0/go.mod h1:8t/X+FVWvk6TaBcsa+UKUBbn7GMtvyBKX30SGl4em6Y=
```
Comment (vincepri) [inline go.sum]: We might have to regenerate some files with the new controller-tools, which brings in new changes.

File: go.sum [hunk 3]
```diff
@@ -254,9 +254,14 @@ k8s.io/utils v0.0.0-20190607212802-c55fbcfc754a h1:2jUDc9gJja832Ftp+QbDV0tVhQHMI
 k8s.io/utils v0.0.0-20190607212802-c55fbcfc754a/go.mod h1:sZAwmy6armz5eXlNoLmJcl4F1QuKu7sr+mFQ0byX7Ew=
 sigs.k8s.io/cluster-api v0.0.0-20190809134526-b8af96f0a42a h1:aDiczQxkpWlxe5yzqm6rJNe5orMuHTt7Q0MqDJOJ/bg=
 sigs.k8s.io/cluster-api v0.0.0-20190809134526-b8af96f0a42a/go.mod h1:CgxuxJy6W8onO8duP2ED+s0kwNxsNWIGEXMnSUsxGcw=
+sigs.k8s.io/cluster-api v0.0.0-20190820130448-9fd8e4cbea0f h1:gL8HQOSMGhMvL2BZNQSuKdXe02se34Uf29Rvq3iOfP8=
+sigs.k8s.io/cluster-api v0.0.0-20190820130448-9fd8e4cbea0f/go.mod h1:zBOsRI6nVKzD6CPmA/3qYpLCckc5FVxMBxhvBxuDNsg=
```
Comment (vincepri) [inline go.sum]: We should run `go mod tidy` or look at why we're getting double entries for these modules

Comment (detiber) [reply go.sum]: I'm trying to work through issues now, but `go mod` is making me angry, `go mod tidy` tries to switch to capi v0.1.9 currently.

Comment (vincepri) [reply go.sum]: We should have a `replace` directive for Cluster API, k8s.io/api and apimachinery, like in CAPA:

```
replace (
	k8s.io/api => k8s.io/api v0.0.0-20190704095032-f4ca3d3bdf1d
	k8s.io/apimachinery => k8s.io/apimachinery v0.0.0-20190704094733-8f6ac2502e51
	sigs.k8s.io/cluster-api => sigs.k8s.io/cluster-api v0.0.0-20190819183132-38edf8b4497e
)
```

File: go.mod [hunk 4]
```diff
@@ -4,16 +4,26 @@ go 1.12
...
 	k8s.io/api v0.0.0-20190409021203-6e4e0e4f393b
 	k8s.io/apimachinery v0.0.0-20190404173353-6a84e37a896d
 	k8s.io/client-go v11.0.1-0.20190409021438-1a26190bd76a+incompatible
 	k8s.io/cluster-bootstrap v0.0.0-20190516232516-d7d78ab2cfe7
 	k8s.io/klog v0.3.3
-	k8s.io/utils v0.0.0-20190607212802-c55fbcfc754a // indirect
-	sigs.k8s.io/cluster-api v0.0.0-20190809134526-b8af96f0a42a
-	sigs.k8s.io/controller-runtime v0.2.0-beta.5
+	k8s.io/utils v0.0.0-20190809000727-6c36bc71fc4a // indirect
+	sigs.k8s.io/cluster-api v0.0.0-00010101000000-000000000000
```
Comment (vincepri) [inline go.mod]: Shouldn't this match the replace directive?

**Comment by detiber** [comment]
@vincepri @chuckha I believe this is sorted out now

**Pull request merged**